### ISSUE TITLE
[BUG?] NullReferenceException when testing API implemented by Refit

### ISSUE THREAD

**Issue by RobThree** [reactiveui/refit#1243]
I'm not sure if this is by-design or a bug, but I ran into an issue recently and thought I'd report it.

For reference: I have [discussed this on stackoverflow question](https://stackoverflow.com/q/69266814/215042).

I was unittesting some REST API code and implemented a test-`HttpMessageHandler` to be able to return any desired (json/html/whatever) response and statuscode (200 OK, 401 Unauthorized, ...). This is the handler in simplified form:

```c#
public class TestHandler : HttpMessageHandler
{
    private readonly HttpResponseMessage _response;

    public TestHandler(HttpStatusCode httpStatusCode)
        => _response = new HttpResponseMessage(httpStatusCode) {
                Content = new StringContent("Yay!")
        };

    protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        => Task.FromResult(_response);
}
```

Then a Refit interface:

```c#
public interface ITest
{
    [Get("/foo/{bar}")]
    Task<string> Test(string bar);
}
```

Finally, a "test":

```c#
var api = RestService.For<ITest>(
  new HttpClient(new TestHandler(HttpStatusCode.OK))
      {
          BaseAddress = new Uri("https://example.com")
      }
  );
  Console.WriteLine(await api.Test("foo").ConfigureAwait(false));
```

This works fine; however, whenever I tried anything other than an OK result I got a `NullReference` exception.

```txt
Object reference not set to an instance of an object.
   at Refit.DefaultApiExceptionFactory.<CreateExceptionAsync>d__4.MoveNext() in /_/Refit/RefitSettings.cs:line 183
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Refit.RequestBuilderImplementation.<>c__DisplayClass14_0`2.<<BuildCancellableTaskFuncForMethod>b__0>d.MoveNext() in /_/Refit/RequestBuilderImplementation.cs:line 313
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at RefitTest.Program.<Main>d__0.MoveNext() in C:\Users\Rob\Source\Repos\RefitTest\RefitTest\Program.cs:line 18
```

Long story short: I finally [found it](https://stackoverflow.com/a/69284500/215042): the returned `HttpResponseMessage` needs a `RequestMessage`:

```c#
    public TestHandler(HttpStatusCode httpStatusCode)
        => _response = new HttpResponseMessage(httpStatusCode) {
                RequestMessage = new(),  // <-- This one here...
                Content = new StringContent("Yay!")
        };
```

Again: I'm not sure if this is a bug, but we tracked it down to [this line](https://github.com/reactiveui/refit/blob/39553f245b929e169809d0be6e77e13a479291a7/Refit/RefitSettings.cs#L183), where `RequestMessage` is null obviously. If anything, maybe a  more explicit null check and more specific exception could be thrown?

**Comment by clairernovotny** [comment]
@RobThree I believe the built in protocol handlers always provide the incoming request message in the response. I'm not 100% sure of the usage contract in the API, but I've never seen it otherwise.

Rather than a code fix, I'd suggest a documentation update to clarify this. Would you mind submitting a PR to mention that in the docs in the right spot?

**Comment by RobThree** [comment]
I don't mind at all. However, what docs are you referring to? The README.md? Those are all the docs I'm aware of? And I can't really find a good place to add it in the (current) docs. There's no explicit unittesting or messagehandling section so I'd have to mention it somewhere as an aside?

**Comment by glennawatson** [comment]
Confirmed and still present on current main. `DefaultApiExceptionFactory.CreateExceptionAsync` force-unwraps `responseMessage.RequestMessage!` (`src/Refit/RefitSettings.cs`), so a test `HttpMessageHandler` that returns an `HttpResponseMessage` without a `RequestMessage` produces a `NullReferenceException` on any non-success path instead of a clear error. The built-in handlers always set `RequestMessage`, but hand-rolled test handlers commonly don't. A more explicit null check / specific exception would be a real improvement. Keeping open as a bug.
